ratis-client/src/main/java/org/apache/ratis/client/impl/DataStreamClientImpl.java

```diff
@@ -153,6 +153,9 @@ private CompletableFuture<DataStreamReply> combineHeader(CompletableFuture<DataS
 
     private CompletableFuture<DataStreamReply> writeAsyncImpl(Object data, long length, Iterable<WriteOption> options) {
       if (isClosed()) {
+        if (data instanceof ByteBuf) {
+          ((ByteBuf) data).release();
+        }
```

Comment:
For the `ByteBuf`, it is only used in `DataStreamManagement` for forwarding a `request` to remote.  The `request` will be released at the end.  So, we should not release here.

Reply:
@szetszwo Thanks for the review. For this part, I'm a little confused.

Can I ask where is this ByteBuf released? (https://github.com/apache/ratis/blob/master/ratis-netty/src/main/java/org/apache/ratis/netty/server/DataStreamManagement.java#L134)

The final release I think is for the retain here. (https://github.com/apache/ratis/blob/master/ratis-netty/src/main/java/org/apache/ratis/netty/NettyDataStreamUtils.java#L165)

Reply:
`DataStreamRequestByteBuf` has a `release()` method: https://github.com/apache/ratis/blob/3efec8d78e143939234748e2d614dbcf5fbba467/ratis-common/src/main/java/org/apache/ratis/datastream/impl/DataStreamRequestByteBuf.java#L70

and the `request` is released in finally-blocks.

Reply:
But here, the request is retained twice, once in "decode", once in remote write. For release only once in finally-blocks, if so, resource will be leak.

Reply:
For this part, since ByteBuf is retained for remote write, the refCount will only decrease on a successful write operation (ctx.writeAndFlush() or ctx.write), but if stream is closed, the write operation is not done, so need to be release here.

Reply:
> Can I ask where is this ByteBuf released? (https://github.com/apache/ratis/blob/master/ratis-netty/src/main/java/org/apache/ratis/netty/server/DataStreamManagement.java#L134)

This one will be released on outbound write.

> The final release I think is for the retain here. (https://github.com/apache/ratis/blob/master/ratis-netty/src/main/java/org/apache/ratis/netty/NettyDataStreamUtils.java#L165)

The final release is for the very first initialize of the ByteBuf. The `retain` in `decode` will be release after the decode itself.

Reply:
> This one will be released on outbound write.

> The final release is for the very first initialize of the ByteBuf. The retain in decode will be release after the decode itself.

The `buf`s seem to be released correctly.  Do you agree?

Reply:
> This one will be released on outbound write.

The ByteBuf in the code should be here. So the retain needs to be released on outbound write. If stream closed, we need to release it manually.

Reply:
@szetszwo I think this `buf` will be release by Netty after outbound write, but if the outbound write be skipped, we should release it manually.

Reply:
@symious sorry that I missed your message earlier.

@xichen01 , thanks for helping out!